### PR TITLE
[CALCITE-2884] Implement JSON_INSERT, JSON_REPLACE, JSON_SET

### PR DESCRIPTION
**Pull request by XuQianJin-Stars**
`JSON_INSERT`(jsondoc, path, val[, path, val] )

`JSON_REPLACE`(jsondoc, _path_, _val_[, path, val] )

`JSON_SET`(jsondoc, path, _val[, path, val] )

Inserts data into a JSON document and returns the result. Returns {{NULL}} if any argument is {{NULL}}. An error occurs if the _{{json_doc}}_ argument is not a valid JSON document or any _{{path}}_ argument is not a valid path expression or contains a  *or {{}}}}*{{{{*}} wildcard.

The path-value pairs are evaluated left to right. The document produced by evaluating one pair becomes the new value against which the next pair is evaluated.

A path-value pair for an existing path in the document is ignored and does not overwrite the existing document value. A path-value pair for a nonexisting path in the document adds the value to the document if the path identifies one of these types of values:
 * A member not present in an existing object. The member is added to the object and associated with the new value.

 * A position past the end of an existing array. The array is extended with the new value. If the existing value is not an array, it is autowrapped as an array, then extended with the new value.

Otherwise, a path-value pair for a nonexisting path in the document is ignored and has no effect.

For a comparison of [JSON_INSERT()](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-insert), [JSON_REPLACE()](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-replace), and [JSON_SET()](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-set), see the discussion of [JSON_SET()](https://dev.mysql.com/doc/refman/5.7/en/json-modification-functions.html#function_json-set).

JSON_INSERT SQL:

```
SELECT JSON_INSERT(v, '$.a', 10, '$.c', '[true, false]') AS c1
 FROM (VALUES ('{ "a": 1, "b": [2, 3]}')) AS t(v);
```

Result:

| c1                                          |
| ------------------------------------------- |
| {"a": 1, "b": [2, 3], "c": "[true, false]"} |

JSON_REPLACE SQL:

```
SELECT JSON_REPLACE(v, '$.a', 10, '$.c', '[true, false]') AS c1
 FROM (VALUES ('{ "a": 1, "b": [2, 3]}')) AS t(v);
```

Result:

| c1                      |
| ----------------------- |
| {"a": 10, "b": [2, 3],} |

JSON_SET SQL:

```
SELECT JSON_INSERT(v, '$.a', 10, '$.c', '[true, false]') AS c1
 FROM (VALUES ('{ "a": 1, "b": [2, 3]}')) AS t(v);
```

Result:

| c1                                           |
| -------------------------------------------- |
| {"a": 10, "b": [2, 3], "c": "[true, false]"} |

### CODE REVIEW

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer I don't know if all the considerations are incomplete. I hope you can review and give me some Suggestions. Thank you very much.
best
qianjin

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -2806,6 +2810,70 @@ private static Integer getJsonDepth(Object o) {
     return depth;
   }
 
+  private static String jsonPairsModify(Object jsonDoc, String type, Object... kvs) {
+    assert kvs.length % 2 == 0;
```
Comment (ritesh-kapoor) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Please help me understand if there were challenges using JSON Path. I'm sorry if it too obvious. https://stackoverflow.com/questions/27244431/how-to-change-values-in-a-json-file-using-xpath-jsonpath-in-java

Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: well，I will change it later.Thank you very much.

Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: hi @ritesh-kapoor Thank you very much. I have seen that jsonpath source code has modified this PR.

**Comment by XuQianJin-Stars** [comment]
hi @michaelmior @zhztheplayer  Thank you very much. Can you review this PR?
best
qianjin

**Comment by ambition119** [comment]
good

**Comment by zhztheplayer** [comment]
> hi @michaelmior @zhztheplayer Thank you very much. Can you review this PR?
> best
> qianjin

I'll help review when I have enough time. Sorry for letting you wait.

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer @ritesh-kapoor Thank you very much，Comments addressed.

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer Thank you very much，Comments addressed.

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 2]
```diff
@@ -95,6 +95,9 @@
   @BaseMessage("Illegal identifier '':''. Was expecting ''VALUE''")
   ExInst<CalciteException> illegalColon();
 
+  @BaseMessage("Illegal identifier '',''. Was expecting ''VALUE''")
+  ExInst<CalciteException> illegalComma();
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java]: I didn't find any usage of this new added method. Am I missing something?

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 3]
```diff
@@ -888,6 +891,15 @@
 
   @BaseMessage("Not a valid input for JSON_STORAGE_SIZE: ''{0}''")
   ExInst<CalciteException> invalidInputForJsonStorageSize(String value);
+
+  @BaseMessage("Not a valid input for JSON_INSERT: jsonDoc: ''{0}'', kvs: ''{1}''")
+  ExInst<CalciteException> invalidInputForJsonInsert(String jsonDoc, String kvs);
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java]: Could you please use "Invalid input" rather than "Not a valid input"? Following are several arguments, not just one.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 4]
```diff
@@ -638,12 +641,120 @@ public static Integer jsonStorageSize(String input) {
   public static Integer jsonStorageSize(JsonValueContext input) {
     try {
       return JSON_PATH_JSON_PROVIDER.getObjectMapper()
-          .writeValueAsBytes(input.obj).length;
+              .writeValueAsBytes(input.obj).length;
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: Indent is not right, please redo the change.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 5]
```diff
@@ -638,12 +641,120 @@ public static Integer jsonStorageSize(String input) {
...
+      } else {
+        throw RESOURCE.validationError(k).ex();
+      }
+    }
+
+    for (Map.Entry<String, Object> entry : map.entrySet()) {
+      String k = entry.getKey();
+      Object v = entry.getValue();
+      switch (type) {
+      case "replace":
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: Any reason not to use an Enum instance to represent "type"?

Comment (zhztheplayer) [reply core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: Also, it would be better we can integrate JSON_REMOVE with a "remove" type.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 6]
```diff
@@ -638,12 +641,120 @@ public static Integer jsonStorageSize(String input) {
...
+    }
+    Object obj = ctx.read(preantPath);
+    if (obj instanceof Map) {
+      ctx.put(preantPath, key, value.toString());
+    } else if (obj instanceof Collection) {
+      ctx.add(preantPath, value);
+    }
+  }
+
+  public static String jsonInsert(Object jsonDoc, Object... kvs) {
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: Please add support for string input. We should support both String and JsonValueContext as the first input argument for such methods. Don't use Object. You can take JSON_REMOVE as an example.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.Locale;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * The <code>JSON_INSERT</code> function.
+ */
+public class SqlJsonInsertFunction extends SqlFunction {
+  public SqlJsonInsertFunction() {
+    super("JSON_INSERT", SqlKind.OTHER_FUNCTION, ReturnTypes.VARCHAR_2000, null,
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java]: Sometimes we return null for JSON functions without necessarily any null input. Unless we are sure that the functions doesn't produce nulls, we should force the return type to nullable in validation. See  JSON_REMOVE as an example.

File: core/src/test/java/org/apache/calcite/test/SqlJsonFunctionsTest.java [hunk 8]
```diff
@@ -521,13 +521,48 @@
 
   @Test public void testJsonRemove() {
     assertJsonRemove(
-        JsonFunctions.jsonValueExpression("{\"a\": 1, \"b\": [2]}"),
-        new String[]{"$.a"},
-        is("{\"b\":[2]}"));
+            JsonFunctions.jsonValueExpression("{\"a\": 1, \"b\": [2]}"),
```
Comment (zhztheplayer) [inline core/src/test/java/org/apache/calcite/test/SqlJsonFunctionsTest.java]: Please make sure you use the right indent: the continuous indention is 4 in Calcite code. So do codes in other new added methods.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  @Override public void unparse(SqlWriter writer, SqlCall call, int leftPrec,
+      int rightPrec) {
+    assert call.operandCount() % 2 == 1;
+    final SqlWriter.Frame frame = writer.startFunCall(getName());
+    call.operand(0).unparse(writer, 0, 0);
+    SqlWriter.Frame listFrame = writer.startList("", "");
+    writer.sep(",");
+    for (int i = 1; i < call.operandCount(); i += 2) {
+      writer.sep(",");
+      writer.keyword("KEY");
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java]: I remember that the keywords `KEY`/`VALUE` are added for JSON_OBJECT(AGG) functions. I don't think MySQL's JSON functions have similar syntax. Could you please double check (the same to JSON_SET, JSON_REPLACE)? Also, if you copied codes from another place, please make sure every line you copied is reasonable for staying at the new place.

File: core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java [hunk 10]
```diff
@@ -4659,6 +4659,49 @@ private void checkNullOperand(SqlTester tester, String op) {
...
+    tester.checkString("json_insert('{ \"a\": 1, \"b\": [2]}', '$.a', 10, '$.c', '[true]')",
+            "{\"a\":1,\"b\":[2],\"c\":\"[true]\"}", "VARCHAR(2000) NOT NULL");
+    tester.checkString("json_insert('{ \"a\": 1, \"b\": [2]}', '$', 10, '$', '[true]')",
+            "{\"a\":1,\"b\":[2]}", "VARCHAR(2000) NOT NULL");
+    tester.checkString("json_insert('{ \"a\": 1, \"b\": [2]}', '$.b[1]', 10)",
+            "{\"a\":1,\"b\":[2,10]}", "VARCHAR(2000) NOT NULL");
+    tester.checkString("json_insert('{\"a\": 1, \"b\": [2, 3, [true]]}', '$.b[3]', 'false')",
+            "{\"a\":1,\"b\":[2,3,[true],\"false\"]}", "VARCHAR(2000) NOT NULL");
+    tester.checkFails("json_insert('{\"a\": 1, \"b\": [2, 3, [true]]}', 'a', 'false')",
+            "(?s).*Not a valid input for.*", true);
```
Comment (zhztheplayer) [inline core/src/test/java/org/apache/calcite/sql/test/SqlOperatorBaseTest.java]: Please add some null checks as well as we have done for other methods (the checks are now always located at the tail of test method body).

**Comment by XuQianJin-Stars** [comment]
hi @zhztheplayer Thank you very much, Comments addressed.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 11]
```diff
@@ -791,6 +912,38 @@ public boolean hasException() {
...
+
+    REMOVE("remove");
+
+    String type;
+
+    JsonModifyMode(String type) {
+      this.type = type;
+    }
+
+    public String getType() {
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: We can remove this method (and the other `getType` method), then use the Enum instance directly.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 12]
```diff
@@ -606,41 +608,160 @@ public static String jsonKeys(JsonPathContext context) {
...
+      throw RESOURCE.invalidInputForJsonStorageSize(Objects.toString(input.obj)).ex();
+    }
+  }
+
+  private static String jsonModify(JsonValueContext jsonDoc, String type, Object... kvs) {
+    JsonModifyMode jsonPairsType = JsonModifyMode.getType(type);
+    if (jsonPairsType != JsonModifyMode.REMOVE) {
+        assert kvs.length % 2 == 0;
+    }
+    String result = "";
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: If I am not wrong. Per operator definition return types of all these methods are nullable. Please either use null or throw error rather than returning empty string when something unexpected occurs.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 13]
```diff
@@ -606,41 +608,160 @@ public static String jsonKeys(JsonPathContext context) {
...
+          break;
+        }
+      }
+    }
+    result = Strings.isNullOrEmpty(result) ?  ctx.jsonString() : result;
+    return result;
+  }
+
+  private static void insertToJson(DocumentContext ctx, String path, Object value) {
+    final String preantPath;
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: s/preant/parent

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 14]
```diff
@@ -606,41 +608,160 @@ public static String jsonKeys(JsonPathContext context) {
...
+    final String key;
+    Integer dotIndex = path.lastIndexOf(".");
+    Integer leftBracketIndex = path.lastIndexOf("[");
+    if (dotIndex.equals(-1) && leftBracketIndex.equals(-1)) {
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: The block of if...else if...else if... is hard to understand. I see you were trying to parse the jsonpath specification by your self. Can we avoid such a implementation? Or could you please explain a bit how it works?

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java [hunk 15]
```diff
@@ -606,41 +608,160 @@ public static String jsonKeys(JsonPathContext context) {
...
+    String result = "";
+    DocumentContext ctx = JsonPath.parse(jsonDoc.obj,
+        Configuration
+            .builder()
+            .options(Option.SUPPRESS_EXCEPTIONS)
+            .jsonProvider(JSON_PATH_JSON_PROVIDER)
+            .mappingProvider(JSON_PATH_MAPPING_PROVIDER)
+            .build());
+
+    if (jsonPairsType == JsonModifyMode.REMOVE) {
```
Comment (zhztheplayer) [inline core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: Can we generalize JsonModifyMode.REMOVE into the switch block below?

File: site/_docs/reference.md [hunk 16]
```diff
@@ -2099,8 +2099,11 @@ semantics.
 | m | JSON_DEPTH(jsonValue)                          | Returns an integer value indicating the depth of a *jsonValue*
 | m | JSON_PRETTY(jsonValue)                         | Returns a pretty-printing of *jsonValue*
 | m | JSON_LENGTH(jsonValue [, path ])               | Returns a integer indicating the length of *jsonValue*
+| m | JSON_INSERT(jsonValue, path, val[, path, val])  | Returns a JSON document insert a data of *jsonValue*, *path*, *val*
 | m | JSON_KEYS(jsonValue [, path ])                 | Returns a string indicating the keys of a JSON *jsonValue*
 | m | JSON_REMOVE(jsonValue, path[, path])           | Removes data from *jsonValue* using a series of *path* expressions and returns the result
+| m | JSON_REPLACE(jsonValue, path, val[, path, val])  | Returns a JSON document replace a data of *jsonValue*, *path*, *val*
+| m | JSON_SET(jsonValue, path, val[, path, val])  | Returns a JSON document set a data of *jsonValue*, *path*, *val*
```
Comment (zhztheplayer) [inline site/_docs/reference.md]: Could you please update the explanation about null cases in the "Note:" section below accordingly?

File: core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties [hunk 17]
```diff
@@ -38,6 +38,7 @@ IllegalBinaryString=Illegal binary string {0}
 IllegalFromEmpty=''FROM'' without operands preceding it is illegal
 IllegalRowExpression=ROW expression encountered in illegal context
 IllegalColon=Illegal identifier '':''. Was expecting ''VALUE''
+IllegalComma=Illegal identifier '',''. Was expecting ''VALUE''
```
Comment (zhztheplayer) [inline core/src/main/resources/org/apache/calcite/runtime/CalciteResource.properties]: Is this line redundant?

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 18]
```diff
@@ -11067,6 +11067,30 @@ private void checkCustomColumnResolving(String table) {
...
+    checkExpType("json_insert('{ \"a\": 1, \"b\": [2]}', '$.a', 10, '$.c', '[true]')",
+            "VARCHAR(2000)");
+    checkFails("select ^json_insert('{\"foo\":\"bar\"}')^",
+            "(?s).*Invalid number of arguments.*");
+  }
+
+  @Test public void testJsonReplace() {
+    checkExp("json_replace('{ \"a\": 1, \"b\": [2]}', '$.a', 10, '$.c', '[true]')");
+    checkExpType("json_replace('{ \"a\": 1, \"b\": [2]}', '$.a', 10, '$.c', '[true]')",
+            "VARCHAR(2000)");
```
Comment (zhztheplayer) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: The indents are still incorrect.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java (hunk 14 above)
Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: > The block of if...else if...else if... is hard to understand. I see you were trying to parse the jsonpath specification by your self. Can we avoid such a implementation? Or could you please explain a bit how it works?

Used to get the parent node

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.Locale;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * The <code>JSON_INSERT</code> function.
+ */
+public class SqlJsonInsertFunction extends SqlFunction {
+  public SqlJsonInsertFunction() {
+    super("JSON_INSERT",
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java]: The 3 new `SqlJsonXXXFunction` you add almost have the same implementation, please abstract a base class for them.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    assert (call.operandCount() >= 3) && (call.operandCount() % 2 == 1);
+  }
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+      boolean throwOnFailure) {
+    final int count = callBinding.getOperandCount();
+    for (int i = 1; i < count; i += 2) {
+      RelDataType nameType = callBinding.getOperandType(i);
+      if (!SqlTypeUtil.inCharFamily(nameType)) {
+        if (throwOnFailure) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java]: `inCharFamiy` or `isCharacter`

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java [hunk 21]
```diff
@@ -1340,6 +1340,15 @@ public boolean argumentMustBeScalar(int ordinal) {
   @Deprecated // to be removed before 2.0
   public static final SqlFunction JSON_STORAGE_SIZE = SqlLibraryOperators.JSON_STORAGE_SIZE;
 
+  @Deprecated // to be removed before 2.0
+  public static final SqlFunction JSON_INSERT = SqlLibraryOperators.JSON_INSERT;
+
+  @Deprecated // to be removed before 2.0
+  public static final SqlFunction JSON_REPLACE = SqlLibraryOperators.JSON_REPLACE;
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: Revert the change.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 22]
```diff
@@ -4023,6 +4023,27 @@ private void checkLiteral2(String expression, String expected) {
     sql(query).ok(expected);
   }
 
+  @Test public void testJsonInsert() {
+    String query = "select json_insert(\"product_name\", '$', 10) from \"product\"";
+    final String expected = "SELECT JSON_INSERT(\"product_name\", '$', 10)\n"
+        + "FROM \"foodmart\".\"product\"";
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: For all the tests, add 2 more kind of value test:

1. null values
2. character that needs to encode/decodes e.g. "\n \" "\t" "\r" ..

File: core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 23]
```diff
@@ -8548,6 +8548,21 @@ public void subTestIntervalSecondFailsValidation() {
         .ok("JSON_STORAGE_SIZE(NULL)");
   }
 
+  @Test public void testJsonInsert() {
+    checkExp("json_insert('{ \"a\": 1, \"b\": [2]}', '$.c', '[true]')",
+        "JSON_INSERT('{ \"a\": 1, \"b\": [2]}', '$.c', '[true]')");
+  }
+
+  @Test public void testJsonReplace() {
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java]: Why we need a SqlParserTest ?

File: site/_docs/reference.md (hunk 16 above)
Comment (danny0405) [reply site/_docs/reference.md]: is `[, path, val]` part means we can have more than one pair of `path` -> `val`, if that is the case, please add a "*" after that.

**Comment by XuQianJin-Stars** [comment]
hi @danny0405 Thank you very much, I will start to modify as soon as possible.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 21 above)
Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: > Revert the change.

Currently, it cannot be temporarily deleted. I will modify the entire JSON function later.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+package org.apache.calcite.sql.fun;
+
+/**
+ * The <code>JSON_INSERT</code> function.
+ */
+public class SqlJsonInsertFunction extends SqlJsonModifyFunction {
+  public SqlJsonInsertFunction(String name) {
+    super(name);
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java]: The sub-classes of `SqlJsonModifyFunction` can all be removed.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 21 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: ???

Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: > 

Because I found that deleting these lines will cause the `testJsonInsert`, `testJsonReplace`, and `testJsonSet` in `RelToSqlConverterTest` to fail even if `sql(query).withMysql ().Ok (expected)`.

Comment (XuQianJin-Stars) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: > ???

Because I found that deleting these lines will cause the `testJsonInsert`, `testJsonReplace`, and `testJsonSet` in `RelToSqlConverterTest` to fail even if `sql(query).withMysql ().Ok (expected)`.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: Then fix that.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java (hunk 22 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: 1. Do not test the 'null' and "\n" in one query
2. put each query and expected pair together instead of putting all the queries together.

File: core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java (hunk 14 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/runtime/JsonFunctions.java]: I'm not familiar with the JSON parse, @zhztheplayer  can you help to review that part ?

**Comment by oleg-smith** [comment]
@zabetak @julianhyde is there a chance to merge this PR? 
Looks like all comments were addressed.

**Comment by julianhyde** [comment]
@oleg-smith The PR looks reasonable but is very old and so needs to rebased. If someone rebases I will merge.

Please move the conversation to https://issues.apache.org/jira/browse/CALCITE-2884. I don't receive most GitHub notifications.

**Comment by oleg-smith** [comment]
@XuQianJin-Stars any chance to rebase it, as it's ready to be merged? I'd really appreciate it!

**Comment by zabetak** [comment]
@oleg-smith I am rebasing now and will try to get this by the end of today.

**Comment by zabetak** [comment]
Closing in favor of https://github.com/apache/calcite/pull/3020. I didn't push the changes here cause I didn't want to rebase a branch that I do not own.
